### PR TITLE
Add a manufactured solution to MPAS-Ocean

### PR DESCRIPTION
**Pull request by sbrus89**
This PR adds the tendencies necessary for a manufactured solution test case in MPAS-Ocean. The details of the case can be found in Section 3.7 of a pre-print by Sid Bishnu et al. https://doi.org/10.22541/essoar.167100170.03833124/v1. 

A convergence study that uses this feature has been added to the polaris package https://github.com/E3SM-Project/polaris/pull/72. 

Note that currently this case shows first order convergence due to the issue #5364. However, corrections to the RK4 substep time forcing yields the expected 2nd order convergence. A follow-on PR will be submitted to address this issue.

[BFB]
[NML]

### CODE REVIEW

**Comment by mark-petersen** [comment]
Thanks! I reviewed the code and it looks correct. I also compiled with gnu and intel.

@sbrus89 can you comment here on how you tested this PR?

**Comment by sbrus89** [comment]
@mark-petersen I tested this with https://github.com/E3SM-Project/polaris/pull/72 with debug flags on Perlmutter. The output plots for that case are below. The sub-optimal convergence is due to https://github.com/E3SM-Project/E3SM/issues/5364

comparison
convergence

**Review by mark-petersen — APPROVED** [review]
@sbrus89 thanks for adding this test case. Approving based on visual inspection and your results.

File: components/mpas-ocean/src/Registry.xml [hunk 1]
```diff
@@ -1533,6 +1533,24 @@
 					possible_values="Any positive non-zero integer. A value of -1 causes this to be overwritten with the configurations vertical level definition."
 		/>
 	</nml_record>
+	<nml_record name="manufactured_solution">
+		<nml_option name="config_use_manufactured_solution" type="logical" default_value=".false."
+					description="This flag includes additional thickness and velocity tendencies nesseary for testing with a manufactured solution."
```
Comment (cbegeman) [inline components/mpas-ocean/src/Registry.xml:1538]:
```suggestion
					description="This flag includes additional thickness and velocity tendencies necessary for testing with a manufactured solution."
```

File: components/mpas-ocean/src/Registry.xml [hunk 2]
```diff
@@ -1533,6 +1533,24 @@
...
+					possible_values=".true. or .false."
+		/>
+		<nml_option name="config_manufactured_solution_wavelength_x" type="real" default_value="2000000.0" units="m"
+					description="Wavelenth of manufactured solution in the x direction"
```
Comment (cbegeman) [inline components/mpas-ocean/src/Registry.xml:1542]:
```suggestion
					description="Wavelength of manufactured solution in the x direction"
```

File: components/mpas-ocean/src/Registry.xml [hunk 3]
```diff
@@ -1533,6 +1533,24 @@
...
+					possible_values="Any positive real number"
+		/>
+		<nml_option name="config_manufactured_solution_wavelength_y" type="real" default_value="2000000.0" units="m"
+					description="Wavelenth of manufactured solution in the y direction"
```
Comment (cbegeman) [inline components/mpas-ocean/src/Registry.xml:1546]:
```suggestion
					description="Wavelength of manufactured solution in the y direction"
```

File: components/mpas-ocean/src/shared/mpas_ocn_manufactured_solution.F [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+         kmax = maxLevelCell(iCell)
+
+         phase = kx*xCell(iCell) + ky*yCell(iCell) - ang_freq*time
+         do k = kmin, kmax
+            tend(k,iCell) = tend(k,iCell) + eta0*(-H0*(kx + ky)*sin(phase) &
+                                                  - ang_freq*cos(phase) &
+                                                  + eta0*(kx + ky)*cos(2.0_RKIND*phase))
+         enddo
+
+         err = 0
```
Comment (cbegeman) [inline components/mpas-ocean/src/shared/mpas_ocn_manufactured_solution.F:121]: Is there a reason why you put this in the iCell loop?

File: components/mpas-ocean/src/shared/mpas_ocn_manufactured_solution.F [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+      type (mpas_pool_type), pointer :: verticalMeshPool
+      real (kind=RKIND), dimension(:,:), pointer :: restingThickness
+
+      if (.not. config_use_manufactured_solution) return
+
+      kx = 2.0_RKIND*pii / config_manufactured_solution_wavelength_x
+      ky = 2.0_RKIND*pii / config_manufactured_solution_wavelength_y
+
+      eta0 = config_manufactured_solution_amplitude
+
```
Comment (cbegeman) [inline components/mpas-ocean/src/shared/mpas_ocn_manufactured_solution.F:224]:
```suggestion

      ! This test case assumes that the restingThickness is horizontally uniform
      ! and that only one vertical level is used so only one set of indices is used here.
```
Something like this

Comment (sbrus89) [reply components/mpas-ocean/src/shared/mpas_ocn_manufactured_solution.F:224]: Thanks for adding this clarification comment!

**Comment by sbrus89** [comment]
@jonbob, just an FYI that this is ready to go. I'll let you run the E3SM namelist scripts, unless you'd like me to take care of it.

**Comment by sbrus89** [comment]
The RK4 fix to correct the sub-optimal convergence is under discussion here: https://github.com/E3SM-Ocean-Discussion/E3SM/pull/55

**Comment by jonbob** [comment]
Thanks @sbrus89 -- I'll run the scripts today and make sure this is good to go

**Comment by jonbob** [comment]
@sbrus89 -- done. I'll mark this as ready to merge

**Comment by sbrus89** [comment]
@jonbob, Thanks very much!

**Comment by jonbob** [comment]
passes:
* SMS_D_Ld3.T62_oQU120.CMPASO-IAF.chrysalis_intel
* ERS_Ld5.T62_oQU120.CMPASO-NYF.chrysalis_intel
* SMS_D_Ld1.ne30pg2_EC30to60E2r2.WCYCL1850.chrysalis_intel.allactive-wcprod

with expected NML DIFFs. Merged to next

**Comment by jonbob** [comment]
merged to master and expected NML DIFFs blessed

**Pull request merged**